proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,278 @@
+A Defi smart contract with two instructions IxA and IxB. Both IxA and IxB issue a rebate.
+IxA is an instruction that places an order on the market which can be used to extract profit.
+IxB is an instruction that just does some bookkeeping like settling funds overall harmless instruction.
+Malicious users then can create a custom smart contract to bypass the application fees where it CPI's IxA
+only if they can extract profit or else they use IxB to issue a rebate for the application fees.
+So DApp developers have to be sure when to do rebates usually white listing and black listing instruction sequence
+would be ideal.
+
+A dapp can break the cpi interface with other dapps if it implements this feature. This is because
+it will require additional account for application fees program to all the instructions which calls 
+`CheckApplicationFees` or `Rebate`, interface also has to add an additional instruction `PayApplicationFees`
+to the correct account.
+
+Overall this feature will incentivise creation of proper transaction and spammers would have to pay
+much more fees reducing congestion in the cluster.
+
+## Security Considerations
+
+User could overpay (more than required by dapp) application fees, if the `CheckApplicationFees` instruction 
+has amount more than the application fee required by dapp.
+
+Denial of service attack for a dapp is possible by flooding the cluster with a lot of transactions write-locking
+an account used by the dapp. This attack is already possible on the network.
+This implementation of application fees does not protect the dapp from denial of service
+attacks. An attacker can always flood the network with the transactions without the `CheckApplicationFees` instruction.
+None of these transactions will be executed successfully but the write lock on the account was taken without any
+payment of application fees.
+
+To solve this kind of attack the application fees should be stored in the ledger but this involves a lot of changes in
+the solana core code. And this kind of attack can only affect one dapp at a time, attacker has to burn a lot of gas 
+fees to sustain for a very long time.
+
+
+## Backwards Compatibility
+
+This feature does not introduce any breaking changes. The transaction without using this feature should
+work as it is.
+To use this feature supermajority of the validators should move to a branch which implements this feature.
+Validators which do not implement this feature cannot replay the blocks with transaction using application fees.
+
+## Mango V4 Usecase
+
+With this feature implemented Mango-V4 will be able to charge users who spam risk-free aribitrage
+or spam liquidations by increasing application fees on perp-markets, token banks
+and mango-user accounts.
+
+#### Perp markets
+
+Application fees on perp liquidations, perp place order, perp cancel, perp consume, perp settle fees.
+Rebates on : successful liquidations, consume events, HFT marketmaking refresh
```

Comment:
Unrelated to the core proposal, but note that this likely reduces the efficiency of liquidations because liquidators need to price in the cost of a failing tx.

Reply:
Noted.